openshift/odo
# Devfile v2 command group impl + Odo catalog list components

Signed-off-by: Maysun J Faisal <maysun.j.faisal@ibm.com>

**What type of PR is this?**
/kind feature

**What does does this PR do / why we need it**:
- Implements complete feature of devfile v2 spec for command groups cdc998972f657dcbc5333fcfdbeef487581e95b3
- Refactored and cleaned up code around devfile parser and validate to be consistent across odo commands a82ca625b691343cc37824e9a938669ccb545e24
- Update odo catalog to list components from new default devfile v2 registry 13d67fcef528742e7c592f09d33d0e68169fb7dc
- Update odo catalog list components to only read index.json entries for name instead of parsing the entire devfile. We also no longer show the `SUPPORTED` column for devfile catalog, see discussion at https://github.com/openshift/odo/issues/3249#issuecomment-641350713 065f19b3fc1f3e08cc1b146af239bc01559a10c9 9da66e99d07ae82b2679216ae466130c71996bdb
- Unit tests 065f19b3fc1f3e08cc1b146af239bc01559a10c9
- Fix odo catalog & odo create bugs 66deec8bfc8db2d38a1d036551d34d3ab78eb84f
- integration tests 6d2a0aa2c011649e84ad2f5ad91363d86b7e48ca

**Which issue(s) this PR fixes**:

Fixes #2938 #3271 #3249 #3280 #3436

**How to test changes / Special notes to the reviewer**:
- `odo catalog list components` (now lists new v2 devfiles)
```
$ odo catalog list components
Odo OpenShift Components:
NAME              PROJECT       TAGS                        SUPPORTED
java              openshift     11,8,latest                 YES
nodejs            openshift     10-SCL,8,8-RHOAR,latest     YES
dotnet            openshift     2.1,2.2,3.0,latest          NO
golang            openshift     1.11.5,latest               NO
httpd             
Odo Devfile Components:
NAME            DESCRIPTION                            REGISTRY
maven           Upstream Maven and OpenJDK 11          DefaultDevfileRegistry
nodejs          Stack with NodeJS 10                   DefaultDevfileRegistry
openLiberty     Open Liberty microservice in Java      DefaultDevfileRegistry
quarkus         Upstream Quarkus with Java+GraalVM     DefaultDevfileRegistry
springBoot      Spring Boot® using Java                DefaultDevfileRegistry
```
- `odo create springBoot mondayspringboot`
- `odo url create`
- `odo push`

To test out the various push scenarios:

1. commands with multiple default will error out
2. commands with multiple default but with `odo push --build-command --run-command` should succeed
3. push without a run group should error out
4. push with flags using the wrong group should error out like `odo push --build-command="a run command of group run"`
